### PR TITLE
[Codegen] Refactor the Codegen

### PR DESCRIPTION
**Pull request by luke-hill**
### 🤔 What's changed?

Change the codegen to be structured neatly and properly. Instead of just tacked into the jsonschema

It will live in its own TLD. It will also be split into the subsequent classes, cleaned up via rubocop and eventually neatened up to remove and redundancies / sub-optimisations.

NB: This isn't aimed to improve the codegen much. I know the way each language interprets the generated items is different. For now this is just aimed as a "1st Pass" at tidying up.

### ⚡️ What's your motivation? 

Tidy up messages codebase. Improve Codegen so that we can read it and optimise it more down the line

### 🏷️ What kind of change is this?

- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### ♻️ Anything particular you want feedback on?

This can be thought of as an "Internal" Breaking change. But given no-one who is a consumer will see this breaking change (As all codegen code will be identical), it can go out as a patch release

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

**Comment by luke-hill** [comment]
@clrudolphi this is about 85% complete at the moment. I've got 2 more batches of things to tidy up, but it's essentially almost done.

Again as mentioned if there's going to be a close time overlap, I'll hold off on merging this until you've got yours in that way I can fix up the conflicts my end.

**Comment by clrudolphi** [comment]
@luke-hill Thanks for the heads-up. I think we're done, just pending final reviews by @mpkorstanje and @gasparnagy 
Will your work finish up in the next day or two? If not, I hope we will slip in ahead of you.

**Comment by luke-hill** [comment]
Ping @clrudolphi - I've only made some minor changes to the dotNET stuff here, but if you want to check it out go ahead

**Comment by clrudolphi** [comment]
> Ping @clrudolphi - I've only made some minor changes to the dotNET stuff here, but if you want to check it out go ahead

Looks good.

File: README.md [hunk 1]
```diff
@@ -32,15 +32,19 @@ Each message in a message stream is of type [Envelope](messages.md#envelope).
...
 - Java
 - Javascript
 - Perl
 - PHP
 - Ruby
 
+There are also 2 subdirectories which have the legacy implementation which was largely driven by protobuf.
+These are currently not implemented and are in the process of being ported over to the JSON schema protocol
+
+- .Net
```
[the remark this answers is not in the document]
Comment (gasparnagy) [reply README.md]: I would say that until we don't have the json report gen ready, there might be smaller changes needed, so I would announce it after we have it ready on Reqnroll side.

**Pull request merged**